SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs

```diff
@@ -139,7 +139,7 @@ static partial class EnumerableExtensions
       ArgumentUtility.CheckNotNull ("predicate", predicate);
       ArgumentUtility.CheckNotNull ("createNoMatchingElementException", createNoMatchingElementException);
 
-      TSource result = default (TSource);
+      TSource result = default (TSource)!;

```

Comment:
unfortunately this is currently not supported by NRT, there are plans to allow
either a T: unconstrained or T?? to signify the default value of an unconstrained type param

Reply:
this means there is a compiler error if we do not set the ! at the end? Even though TSource can be nullable?

Reply:
We need to add a test that shows that the sequence can contain null values and everything still compiles. Plus tests to show that Single also works with a sequence of valuetype and nullable value type

Reply:
> this means there is a compiler error if we do not set the ! at the end? Even though TSource can be nullable?

`default(TSource)` is `null` for a reference type, nullable or not. `TSource` could be a non-nullable reference type, which would get assigned `null`, that's why it does not compile without `!`.

Reply:
> We need to add a test that shows that the sequence can contain null values and everything still compiles. Plus tests to show that Single also works with a sequence of valuetype and nullable value type

As `TSource` is unconstrained there is no difference to the code before NRT, so I do not see a value in adding tests (they would just test the compiler, right?), as this is no runtime feature. Am I missing something?

Reply:
They test that the syntax works for all the specified cases. Given there is some weird magic involved, we should make sure that all supported scenarios compile and read syntactically clean on the caller side.

Reply:
> 
> 
> > this means there is a compiler error if we do not set the ! at the end? Even though TSource can be nullable?
> 
> `default(TSource)` is `null` for a reference type, nullable or not. `TSource` could be a non-nullable reference type, which would get assigned `null`, that's why it does not compile without `!`.

So it doesn't affect the typing of TSource, it's just an "ignore this and move on"? I was always under the impression it's the other way around, saying that a possibly null-value won't be null and thus it's okay to assign to a not-nullable variable/parameter/etc.

Reply:
The actual type of `default(TSource)` cant be represented in current C#, because for value types it would be `TSource` and for ref types `TSource?`. The null forgiving operator converts the expression to type `TSource` regardless of actual nullability. 
There is currently a draft for the syntax ``TSource?? result = default(TSource);` for unconstrained type parameters.

Reply:
Yes. 

I just checked the code base and we've got about 15 usages of generic default values. For now, my suggestion stands if it wouldn't be cleaner to just disable NRT for the assignment line using a #enable / #disable instead of using the null-forgiving operator. It just feels cleaner/easier to understand to openly express that NRT doesn't work in this generic situation. null-forgiving is okay in test code when we need to initialize fields that we want to declare as not-nullable, but here in the production code, this particualr semantics would be turned on it's head: the null-forgiving operator would be used to assign a potentially null value as "not null" to a variable of undetermined null-ability. the casual reader would then see both types of usage and step into this exact problem I fell into.

Reply:
Did we decide to use `!` everywhere to be consistent with corefx etc. ?

Reply:
Yes.